pkg/client/client_test.go

```diff
@@ -2199,1464 +2205,1765 @@ U5wwSivyi7vmegHKmblOzNVKA5qPO8zWzqBC
+			Spec: appsv1.DeploymentSpec{
+				Replicas: ptr.To[int32](2),
+				Selector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{"foo": "bar"},
+				},
+				Template: corev1.PodTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{Labels: map[string]string{"foo": "bar"}},
+					Spec:       corev1.PodSpec{Containers: []corev1.Container{{Name: "nginx", Image: "nginx"}}},
+				},
+			},
+		}
+	})
 
-				By("listing the 2 remaining deployments when previous continuation token is used without a limit")
-				deps = &appsv1.DeploymentList{}
-				err = cl.List(context.Background(), deps,
-					client.Continue(continueToken),
-				)
-				Expect(err).NotTo(HaveOccurred())
+	AfterEach(func() {
+		deleteDeployment(ctx, clientset, dep, ns)
+	})
 
-				Expect(deps.Items).To(HaveLen(2))
-				Expect(deps.Continue).To(BeEmpty())
-				Expect(deps.Items[0].Name).To(Equal(dep3.Name))
-				Expect(deps.Items[1].Name).To(Equal(dep4.Name))
-			})
+	Context("with structured objects", func() {
+		It("should fetch collection of objects", func() {
+			By("creating an initial object")
+			dep, err := clientset.AppsV1().Deployments(ns).Create(ctx, dep, metav1.CreateOptions{})
+			Expect(err).NotTo(HaveOccurred())
 
-			PIt("should fail if the object doesn't have meta", func() {
+			cl, err := client.New(cfg, client.Options{})
+			Expect(err).NotTo(HaveOccurred())
 
-			})
+			var logs []string
+			ctx := pkglog.IntoContext(context.Background(), funcr.New(func(_, msg string) {
+				logs = append(logs, msg)
+			}, funcr.Options{Verbosity: 8})) // Increase log level so we can use to determine if watchList and/or list was called.
 
-			PIt("should fail if the object cannot be mapped to a GVK", func() {
+			By("listing all objects of that type in the cluster")
+			deps := &appsv1.DeploymentList{}
+			Expect(cl.List(ctx, deps)).To(Succeed())
 
-			})
+			Expect(deps.Items).NotTo(BeEmpty())
+			hasDep := false
+			for _, item := range deps.Items {
+				if item.Name == dep.Name && item.Namespace == dep.Namespace {
+					hasDep = true
+					break
+				}
+			}
+			Expect(hasDep).To(BeTrue())
+
+			switch {
+			case watchListEnabled && watchListClientEnabled:
+				// Expect successful watchList
+				// We expect that no response bodies are logged
+				Expect(logs).ToNot(ContainElement(MatchRegexp("Response Body.*DeploymentList")))
+			case !watchListEnabled && watchListClientEnabled:
+				// Expect failed watchList + successful fallback to list
+				Expect(logs).To(ContainElements(
+					"\"msg\"=\"Warning: The watchlist request for deployments ended with an error, " +
+						"falling back to the standard LIST semantics\" " +
+						"\"error\"=\"ListOptions.meta.k8s.io \\\"\\\" is invalid: sendInitialEvents: " +
+						"Forbidden: sendInitialEvents is forbidden for watch unless the WatchList feature gate is enabled\""))
+				// Fallback list logs response body
+				Expect(logs).To(ContainElement(MatchRegexp("Response Body.*DeploymentList")))
```

Comment:
Its safe to assume this is absent if watchlist is used because then we never get an `ObjectList` type but only event types?

Reply:
A working watchList call logs no ResponseBodies at all.

WatchList: true WatchListClient: true
![image](https://github.com/user-attachments/assets/f7f7f117-bbe3-4ac9-9714-61e1364b68f2)

Failing watchList + successful list looks like this

WatchList: false WatchListClient: true
![image](https://github.com/user-attachments/assets/ce8cc786-c0ec-4a63-8001-dc09971d74fb)

I know it's not super great to rely on these logs and the tests might break when we bump, but it's the best thing I found to figure out what if watchList and/or list are called

Reply:
Yeah, I am okay with this. An alternate approach would be to inject a custom http roundtripper and look at the requests

Reply:
Alrighty. I'll keep it as is for now. Let's consider your alternative if this turns out to be too brittle